// modulorj4/index.js
class Cliente {
    #cpf //privado
    constructor(nome, cpf, endereco){
        this.nome = nome
        this.endereco = endereco
        this.#cpf = cpf
        this.telefones = new Set()
        
    }
    //get
    get getCpf(){
        return this.#cpf
    }

    get upperNome(){
        return this.nome.toUpperCase()
    }
    get lowerNome(){
        return this.nome.toLowerCase()
    }

    get upperEndereco(){
        return this.endereco.toUpperCase()
    }
    get lowerEndereco(){
        return this.endereco.toLowerCase()
    }

    addTelefone(telefone){
        if(telefone instanceof Telefone){
            this.telefones.add(telefone)
        }
    }
}

class Telefone {
    constructor(ddd, numero){
        this.ddd = ddd
        this.numero = numero
    }
}

class Endereco {
    constructor(estado, cidade, rua, numero){
        this.estado = estado
        this.cidade = cidade
        this.rua = rua
        this.numero = numero
    }

    get upperEstado(){
        return this.estado.toUpperCase()
    }
    get lowerEstado(){
        return this.estado.toLowerCase()
    }

    get upperCidade(){
        return this.cidade.toUpperCase()
    }
    get lowerCidade(){
        return this.cidade.toLowerCase()
    }

    get upperRua(){
        return this.rua.toUpperCase()
    }
    get lowerRua(){
        return this.rua.toLowerCase()
    }
}

class Empresa {
    #cnpj //privado
    constructor(razaoSocial, nomeFantasia, cnpj, endereco){
        this.endereco = endereco
        this.nomeFantasia = nomeFantasia
        this.razaoSocial = razaoSocial
        this.#cnpj = cnpj
        this.clientes = new Set()
        this.telefones = new Set()
    }

    get getCnpj(){
        return this.#cnpj
    }

    get upperEndereco(){
        return this.endereco.toUpperCase()
    }
    get lowerEndereco(){
        return this.endereco.toLowerCase()
    }

    get upperNomeFantasia(){
        return this.nomeFantasia.toUpperCase()
    }
    get lowerNomeFantasia(){
        return this.nomeFantasia.toLowerCase()
    }

    get upperRazaoSocial(){
        return this.razaoSocial.toUpperCase()
    }
    get lowerRazaoSocial(){
        return this.razaoSocial.toLowerCase()
    }

    get upperClientes(){
        return this.clientes.toUpperCase()
    }
    get lowerClientes(){
        return this.clientes.toLowerCase()
    }

    addCliente(cliente){
        if (cliente instanceof Cliente) {
            this.clientes.add(cliente);
        }
    }

    descricao(){
        //saida
    let resultado = `Razão Social: ${this.razaoSocial}
        Nome Fantasia: ${this.nomeFantasia}
        `
        for(let cliente of this.clientes){
            let telefones = ''

            for(let telefone of cliente.telefones){
                telefones += `\nDDD: ${telefone.ddd} .......... Número: ${telefone.numero}` 
            }

            resultado += `
--------------------
Nome: ${cliente.nome}
Estado: ${cliente.endereco.estado} , Cidade: ${cliente.endereco.cidade} , Rua: ${cliente.endereco.rua} , Número: ${cliente.endereco.numero} ${telefones}
`
        }
        return resultado;
    }
}

export {Cliente, Telefone, Endereco, Empresa}
